### ISSUE TITLE
Converting from Pull to Stream should introduce a scope

### ISSUE THREAD

**Issue by mpilquist** [functional-streams-for-scala/fs2#1154]
Consider this example:

```scala
val s = Stream(1, 2, 3).onFinalize(IO(println("finalized")))
val prg = s.take(1).repeat.take(10).append(Stream.eval(IO(Thread.sleep(5000))))
```

When `prg` is run, the finalizers are all delayed until the overall stream terminates. The root cause of this is `take(1)`, which invokes uncons once and then discards the tail stream. Because the tail stream is discarded, the finalizer for the resource associated with the `onFinalize` call is never explicitly released.

Manually inserting a scope after `take(1)` fixes this issue. Specifically:

```
val prg2 = s.take(1).scope.repeat.take(10).append(Stream.eval(IO(Thread.sleep(5000))))
```

In 0.9, *all* pull-to-scope conversions inserted a scope. We removed this in 0.10 b/c we thought it was unnecessary but in hindsight, I believe this was a mistake. I suspect this is the root of many of the issues exposed in #1151.

**Comment by SystemFw** [comment]
I can't remember if there were any other reasons for not having a scope after every Pull-to-Stream apart from thinking it wasn't necessary

**Comment by mpilquist** [comment]
I added `streamNoScope` back in October: https://github.com/functional-streams-for-scala/fs2/commit/a3019678b086b5554ede9f331515c54b8ee47285

This bugged me at the time, and I really liked the idea of zero cost from going between `Stream` and `Pull` -- just two different monads for the same underlying data structure.

I then removed it in January: https://github.com/functional-streams-for-scala/fs2/commit/47e41dbbea554fa391ff51d48dfd926e8c0c3822

**Comment by mpilquist** [comment]
Fixed by #1155
